## Overview


Classificaation Problem:

Algorithms Used:

- RandomForest wih MLLib. Reference: https://jarrettmeyer.com/2017/05/04/random-forests-with-pyspark
- Decision Tree with MLLib. Reference: https://spark.apache.org/docs/2.2.0/mllib-decision-tree.html
- Logistic Regression with MLLib. Reference: https://spark.apache.org/docs/2.3.0/mllib-linear-methods.html#logistic-regression
- Multilayered Perceptron with Spark ML. Reference: https://spark.apache.org/docs/2.3.0/ml-classification-regression.html#multilayer-perceptron-classifier

## Loading the file into dataframe

In [3]:
## LOADING FILES AND DISPLAYING THE TABLE
# File location and type
file_location = "/FileStore/tables/creditcard.csv"
file_type = "csv"

# CSV options
infer_schema = "true"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

display(df)

Time V1 V2 V3 V4 V5 V6 V7 V8 V9 V10 V11 V12 V13 V14 V15 V16 V17 V18 V19 V20 V21 V22 V23 V24 V25 V26 V27 V28 Amount Class 0 -1.359807134 -0.072781173 2.536346738 1.378155224 -0.33832077 0.462387778 0.239598554 0.098697901 0.36378697 0.090794172 -0.551599533 -0.617800856 -0.991389847 -0.311169354 1.468176972 -0.470400525 0.207971242 0.02579058 0.40399296 0.251412098 -0.018306778 0.277837576 -0.11047391 0.066928075 0.128539358 -0.189114844 0.133558377 -0.021053053 149.62 0 0 1.191857111 0.266150712 0.166480113 0.448154078 0.060017649 -0.082360809 -0.078802983 0.085101655 -0.255425128 -0.166974414 1.612726661 1.065235311 0.489095016 -0.143772296 0.635558093 0.463917041 -0.114804663 -0.18336127 -0.145783041 -0.069083135 -0.225775248 -0.638671953 0.101288021 -0.339846476 0.167170404 0.125894532 -0.008983099 0.014724169 2.69 0 1 -1.358354062 -1.340163075 1.773209343 0.379779593 -0.503198133 1.800499381 0.791460956 0.247675787 -1.514654323 0.207642865 0.624501459 0.066083685 0.717292731 -0.165945923 2.345864949 -2.890083194 1.109969379 -0.121359313 -2.261857095 0.524979725 0.247998153 0.771679402 0.909412262 -0.689280956 -0.327641834 -0.139096572 -0.055352794 -0.059751841 378.66 0 1 -0.966271712 -0.185226008 1.79299334 -0.863291275 -0.01030888 1.247203168 0.23760894 0.377435875 -1.387024063 -0.054951922 -0.226487264 0.178228226 0.50775687 -0.287923745 -0.631418118 -1.059647245 -0.684092786 1.965775003 -1.23262197 -0.208037781 -0.108300452 0.005273597 -0.190320519 -1.175575332 0.647376035 -0.221928844 0.062722849 0.061457629 123.5 0 2 -1.158233093 0.877736755 1.548717847 0.403033934 -0.407193377 0.095921462 0.592940745 -0.270532677 0.817739308 0.753074432 -0.822842878 0.53819555 1.345851593 -1.119669835 0.17512113 -0.451449183 -0.237033239 -0.038194787 0.803486925 0.40854236 -0.009430697 0.798278495 -0.13745808 0.141266984 -0.206009588 0.502292224 0.21942223 0.215153147 69.99 0 2 -0.425965884 0.960523045 1.141109342 -0.16825208 0.420986881 -0.029727552 0.476200949 0.260314333 -0.568671376 -0.371407197 1.34126198 0.359893837 -0.358090653 -0.1371337 0.517616807 0.401725896 -0.058132823 0.068653149 -0.033193788 0.084967672 -0.208253515 -0.559824796 -0.026397668 -0.371426583 -0.232793817 0.105914779 0.253844225 0.081080257 3.67 0 4 1.229657635 0.141003507 0.045370774 1.202612737 0.191880989 0.272708123 -0.005159003 0.08121294 0.464959995 -0.099254321 -1.416907243 -0.153825826 -0.751062716 0.167371963 0.050143594 -0.443586798 0.002820512 -0.61198734 -0.045575045 -0.219632553 -0.167716266 -0.270709726 -0.154103787 -0.780055415 0.750136936 -0.257236846 0.03450743 0.005167769 4.99 0 7 -0.644269442 1.417963545 1.074380376 -0.492199018 0.948934095 0.428118463 1.120631358 -3.807864239 0.615374731 1.249376178 -0.619467796 0.291474353 1.757964214 -1.32386522 0.686132504 -0.076126999 -1.222127345 -0.35822157 0.324504731 -0.156741852 1.94346534 -1.01545471 0.05750353 -0.649709006 -0.415266566 -0.051634297 -1.206921081 -1.085339188 40.8 0 7 -0.894286082 0.286157196 -0.113192213 -0.27152613 2.66959866 3.721818061 0.370145128 0.851084443 -0.392047587 -0.410430433 -0.705116587 -0.110452262 -0.286253632 0.07435536 -0.32878305 -0.210077268 -0.499767969 0.118764861 0.570328167 0.052735669 -0.0734251 -0.268091632 -0.20423267 1.011591802 0.37320468 -0.384157308 0.011747356 0.14240433 93.2 0 9 -0.338261752 1.119593376 1.044366552 -0.222187277 0.499360806 -0.246761101 0.651583206 0.069538587 -0.736727316 -0.366845639 1.017614468 0.83638957 1.006843514 -0.443522817 0.150219101 0.739452777 -0.540979922 0.47667726 0.451772964 0.203711455 -0.246913937 -0.633752642 -0.120794084 -0.385049925 -0.069733046 0.094198834 0.246219305 0.083075649 3.68 0 10 1.449043781 -1.176338825 0.913859833 -1.375666655 -1.971383165 -0.629152139 -1.423235601 0.048455888 -1.720408393 1.626659058 1.19964395 -0.671439778 -0.513947153 -0.095045045 0.230930409 0.031967467 0.253414716 0.854343814 -0.221365414 -0.387226474 -0.009301897 0.313894411 0.027740158 0.500512287 0.251367359 -0.129477954

## RandomForest Classifier

In [5]:
RANDOM_SEED = 13579
TRAINING_DATA_RATIO = 0.7

RF_NUM_TREES = 3
RF_MAX_DEPTH = 4
RF_MAX_BINS = 32

In [6]:
## Data Preprocessing and Test Train Split
from pyspark.mllib.linalg import Vectors
from pyspark.mllib.regression import LabeledPoint

transformed_df = df.rdd.map(lambda row: LabeledPoint(row[-1], Vectors.dense(row[0:-1])))

splits = [TRAINING_DATA_RATIO, 1.0 - TRAINING_DATA_RATIO]
training_data, test_data = transformed_df.randomSplit(splits, RANDOM_SEED)

print("Number of training set rows: %d" % training_data.count())
print("Number of test set rows: %d" % test_data.count())

Number of training set rows: 168447
Number of test set rows: 72048

In [7]:
## Training RandomForest Model
from pyspark.mllib.tree import RandomForest

model_rf = RandomForest.trainClassifier(training_data, numClasses=2, categoricalFeaturesInfo={}, \
    numTrees=RF_NUM_TREES, featureSubsetStrategy="auto", impurity="gini", \
    maxDepth=RF_MAX_DEPTH, maxBins=RF_MAX_BINS, seed=RANDOM_SEED)

In [8]:
## Calculating the accuracy for the RandomForest Model
predictions_rf = model_rf.predict(test_data.map(lambda x: x.features))
labels_and_predictions_rf = test_data.map(lambda x: x.label).zip(predictions_rf)
acc_rf = labels_and_predictions_rf.filter(lambda x: x[0] == x[1]).count() / float(test_data.count())
print("Model accuracy: %.3f%%" % (acc_rf * 100))

Model accuracy: 99.914%

In [9]:
from pyspark.mllib.evaluation import MulticlassMetrics

metrics_rf = MulticlassMetrics(labels_and_predictions_rf)
precision_rf = metrics_rf.precision()
recall_rf = metrics_rf.recall()
fMeasure_rf = metrics_rf.fMeasure()
print("Model precision: %f, recall: %f and f1 score:%f " %((precision_rf* 100), (recall_rf* 100), (fMeasure_rf* 100)))

Model precision: 99.913946, recall: 99.913946 and f1 score:99.913946

In [10]:
## Evaluating the metrics for the RandomForest Model
from pyspark.mllib.evaluation import BinaryClassificationMetrics

metrics_rf = BinaryClassificationMetrics(labels_and_predictions_rf)
print("Area under Precision/Recall (PR) curve: %.f" % (metrics_rf.areaUnderPR * 100))
print("Area under Receiver Operating Characteristic (ROC) curve: %.3f" % (metrics_rf.areaUnderROC * 100))

Area under Precision/Recall (PR) curve: 57
Area under Receiver Operating Characteristic (ROC) curve: 92.823

## Decision Tree Classifier

In [12]:
## Training Decision Tree Model
from pyspark.mllib.tree import DecisionTree, DecisionTreeModel
from pyspark.mllib.util import MLUtils

model_dt = DecisionTree.trainClassifier(training_data, numClasses=2, categoricalFeaturesInfo={},
                                     impurity='gini', maxDepth=RF_MAX_DEPTH, maxBins=RF_MAX_BINS)

In [13]:
## Calculating the accuracy for the Decision Tree Model
predictions_dt = model_dt.predict(test_data.map(lambda x: x.features))
labels_and_predictions_dt = test_data.map(lambda x: x.label).zip(predictions_dt)
acc_dt = labels_and_predictions_dt.filter(lambda x: x[0] == x[1]).count() / float(test_data.count())
print("Model accuracy: %.3f%%" % (acc_dt * 100))

Model accuracy: 99.928%

In [14]:
from pyspark.mllib.evaluation import MulticlassMetrics

metrics_dt = MulticlassMetrics(labels_and_predictions_dt)
precision_dt = metrics_dt.precision()
recall_dt = metrics_dt.recall()
fMeasure_dt = metrics_dt.fMeasure()
print("Model precision: %f, recall: %f and f1 score:%f " %((precision_dt* 100), (recall_dt* 100), (fMeasure_dt* 100)))

Model precision: 99.927826, recall: 99.927826 and f1 score:99.927826

In [15]:
## Evaluating the metrics for the Decision Tree Model
metrics_dt = BinaryClassificationMetrics(labels_and_predictions_dt)
print("Area under Precision/Recall (PR) curve: %.f" % (metrics_dt.areaUnderPR * 100))
print("Area under Receiver Operating Characteristic (ROC) curve: %.3f" % (metrics_dt.areaUnderROC * 100))

Area under Precision/Recall (PR) curve: 69
Area under Receiver Operating Characteristic (ROC) curve: 91.004

## Logistic Regression Classifier

In [17]:
## Training Logistic Regression Model
from pyspark.mllib.classification import LogisticRegressionModel, LogisticRegressionWithLBFGS
model_lr = LogisticRegressionWithLBFGS.train(training_data)

In [18]:
## Calculating the accuracy for the Logistic Regression Model
predictions_lr = model_lr.predict(test_data.map(lambda x: x.features))
labels_and_predictions_lr = test_data.map(lambda x: x.label).zip(predictions_lr)
acc_lr = labels_and_predictions_lr.filter(lambda x: x[0] == x[1]).count() / float(test_data.count())
print("Model accuracy: %.3f%%" % (acc_lr * 100))

Model accuracy: 99.854%

In [19]:
from pyspark.mllib.evaluation import MulticlassMetrics

metrics_lr = MulticlassMetrics(labels_and_predictions_rf)
precision_lr = metrics_lr.precision()
recall_lr = metrics_lr.recall()
fMeasure_lr = metrics_lr.fMeasure()

print("Model precision: %f, recall: %f and f1 score:%f " %((precision_lr* 100), (recall_lr* 100), (fMeasure_lr* 100)))

Model precision: 99.913946, recall: 99.913946 and f1 score:99.913946

### MultiLayered Perceptron Classifier

In [21]:
from pyspark.ml.linalg import DenseVector

input_data = df.rdd.map(lambda x: (x[-1], DenseVector(x[0:-1])))

df = spark.createDataFrame(input_data, ["label", "features"])

In [22]:
training_data1, test_data1 = df.randomSplit([.8,.2],seed=1234)

print("Number of training set rows: %d" % training_data1.count())
print("Number of test set rows: %d" % test_data1.count())

Number of training set rows: 192334
Number of test set rows: 48161

In [23]:
from pyspark.ml.classification import MultilayerPerceptronClassifier
from pyspark.ml.evaluation import MulticlassClassificationEvaluator

layers = [30, 5, 4, 2]
trainer = MultilayerPerceptronClassifier(maxIter=100, layers=layers, blockSize=128, seed=1234)

model_p = trainer.fit(training_data1)

In [24]:
result_p = model_p.transform(test_data1)
predictionAndLabels = result_p.select("prediction", "label")
evaluator = MulticlassClassificationEvaluator(metricName="accuracy")

print("Test set accuracy = " + str(evaluator.evaluate(predictionAndLabels) * 100))

Test set accuracy = 99.82973775461474

In [25]:
precision_p = MulticlassClassificationEvaluator(metricName="weightedPrecision")
recall_p = MulticlassClassificationEvaluator(metricName="weightedRecall")
fScore_p = MulticlassClassificationEvaluator(metricName="f1")

print("Precision = " + str(precision_p.evaluate(predictionAndLabels) * 100))
print("Recall = " + str(recall_p.evaluate(predictionAndLabels) * 100))
print("f1 Score = " + str(fScore_p.evaluate(predictionAndLabels) * 100))

Precision = 99.65976540155151
Recall = 99.82973775461474
f1 Score = 99.74467916675233

### Results

Algorithm/Metrics| Precision | Recall | F1-Score | Accuracy 
--- | --- | --- | --- | --- 
RandomForest | 0.9991|0.9991|0.9991|99.914 
Desicion Tree | 0.9992 |0.9992|0.9992|99.928
Logistic Regression | 0.9991 | 0.9991 | 0.9991 | 99.85
Multilayered Perceptron| 0.9965 | 0.9982 | 0.9974 | 99.82